challenges/back-end-challenge/flask_app.py

```diff
@@ -0,0 +1,52 @@
+
+# A very simple Flask Hello World app for you to get started with...
+
+from flask import Flask, request, Response
+import json
+from os import path
+
+app = Flask(__name__)
+
+sorted_data = []
+
+@app.route('/data', methods=["GET","POST","PATCH"])
+def hello_world():
+
+    try:
+        if request.method == "POST":
+
+            data = request.json
+            if data is None or len(data) != 500:
+                return Response("Exactly 500 numbers are required, please try again.", status=400)
+            data.sort()
+            with open('numbers.json', 'w') as f:
+                json.dump(data, f)
+            return "Data captured successfully! Use GET request to receive sorted list!"
+        elif request.method == "GET":
+
+            if path.exists("numbers.json"):
+                with open('numbers.json', 'r') as f:
+                    data = json.load(f)
+                    return str(data)
+            else:
+                return Response("Please send a list of 500 numbers via POST request first.", status=400)
+        elif request.method == "PATCH":
+
+            if path.exists("numbers.json"):
+                number = request.json
+                if number is None or len(number) != 1:
+                    return Response("Exactly 1 number is required, please try again.", status=400)
+                with open('numbers.json', 'r') as f:
+                    data = json.load(f)
+                data.extend(number)
+                data.sort()
+                with open('numbers.json', 'w') as f:
+                    json.dump(data, f)
+                return str(data)
+            else:
+                return Response("Please send a list of 500 numbers via POST request first.", status=400)
+
+    except Exception as e:
+        return Response(str(e), status=400)
+
+
```

Comment:
If we were receiving a large number of requests to this service, what would be your largest bottleneck for performance?

Reply:
Sorting the numbers takes the most amount of time in our API. Using appropriate sorting algorithms according to the input will make our API more efficient.

Reply:
Do you think sorting a list of 500 numbers in memory would take more or less time than the disk/File IO you are performing using `with open`?

Reply:
Since we are sorting only 500 numbers here, disk IO operations will be costlier than sorting. If we consider 'n' (and not a constant number, 500) sorting would take more time with time complexity O(nlogn).